### PR TITLE
Implement ReadOnlyQuestionService

### PR DESCRIPTION
**Pull request by AWBaumann**
### Description
Implementing all of the read only functions for QuestionService wrapped in a ReadOnlyQuestionService.

### Checklist
- [X] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
[Fixes] #<issue_number>

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java [hunk 1]
```diff
@@ -109,6 +108,19 @@ public QuestionType getQuestionType() {
     return ImmutableMap.of("text", ScalarType.STRING);
   }
 
+  /** Get a map of scalars stored by this question definition. */
+  @JsonIgnore
+  public ImmutableMap<String, ScalarType> getScalars(boolean includeFullPath) {
+    ImmutableMap<String, ScalarType> scalars = this.getScalars();
+    if (includeFullPath) {
+      ImmutableMap.Builder<String, ScalarType> ret = new ImmutableMap.Builder<String, ScalarType>();
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:116]: not sure if this exists but maybe you could use a bistream. this is pseudocode...

`BiStream.from(scalars).mapKeys(k -> this.getPath() + "." + k).toMap();`

Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:116]: it do exist https://google.github.io/mug/apidocs/com/google/mu/util/stream/BiStream.html

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java [hunk 2]
```diff
@@ -71,11 +71,10 @@ public String getDescription() {
   }
 
   /** Get the question text for the given locale. */
-  public String getQuestionText(Locale locale) {
+  public String getQuestionText(Locale locale) throws RuntimeException {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:74]: nit: runtime exceptions don't need to be mentioned in the method signature

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java [hunk 3]
```diff
@@ -85,7 +84,7 @@ public String getQuestionText(Locale locale) {
   }
 
   /** Get the question help text for the given locale. */
-  public String getQuestionHelpText(Locale locale) {
+  public String getQuestionHelpText(Locale locale) throws RuntimeException {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:87]: nit: runtime exceptions don't need to be mentioned in the method signature

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      throws InvalidPathException {
+    PathType pathType = this.getPathType(pathString);
+    if (pathType == PathType.QUESTION) {
+      return questions.get(pathString).getScalars(/* includeFullPath = */ true);
+    } else if (pathType == PathType.SCALAR) {
+      ScalarType scalarType = scalars.get(pathString);
+      return ImmutableMap.of(pathString, scalarType);
+    }
+    throw new InvalidPathException(pathString);
+  }
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:53]: Issue (blocking): do not use `==` for comparing enum types.

See https://engdoc.corp.google.com/eng/doc/devguide/java/practices/enums.md?cl=head#equals

Suggestion: use a `switch` here and L63

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  public PathType getPathType(String pathString) {
+    return questions.containsKey(pathString)
+        ? PathType.QUESTION
+        : scalars.containsKey(pathString) ? PathType.SCALAR : PathType.NONE;
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:58]: Issue (non-blocking): past discussion has resulted in a consensus that nested ternaries are hard to read.

Suggestion: refactor to

```java
if (question.containsKey(pathString)) {
  return PathType.QUESTION;
} else if (scalars.containsKey(pathString)) {
  return PathType.SCALAR;
}

return PathType.NONE;
```

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
+package services.question;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Locale;
+
+public class ReadOnlyQuestionServiceImpl implements ReadOnlyQuestionService {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:7]: Suggestion (non-blocking): add `final`

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java [hunk 7]
```diff
@@ -109,6 +108,19 @@ public QuestionType getQuestionType() {
     return ImmutableMap.of("text", ScalarType.STRING);
   }
 
+  /** Get a map of scalars stored by this question definition. */
+  @JsonIgnore
+  public ImmutableMap<String, ScalarType> getScalars(boolean includeFullPath) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:113]: Issue (non-blocking): avoid using boolean params in public methods

see https://engdoc.corp.google.com/eng/doc/devguide/java/practices/boolean-parameters.md?cl=head

**Review by bion — APPROVED** [review]

**Pull request merged**